### ISSUE TITLE
Linking issue when building Python, TCL, and R wrappers on Arch Linux

### ISSUE THREAD

**Issue by m-pilia** [SimpleITK/SimpleITK#1341]
Hi! I am trying to package version 2.0.2 of SimpleITK for Arch Linux ([PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=simpleitk) on the Arch User Repository package) and I get a linking issue when building the TCL and R wrappers (full configure and build output: [build_output_tcl_R.txt](https://github.com/SimpleITK/SimpleITK/files/6056778/build_output_tcl_R.txt), [build_output_py.txt](https://github.com/SimpleITK/SimpleITK/files/6056777/build_output_py.txt)).

From the errors, it seems there is something wrong with the library path. I used to build the previous versions in the same fashion without issues, so I was wondering if there is something wrong with the way I configure the build, or if this might be a regression in the build configuration. Any help or suggestion would be appreciated, thanks!

I am building SimpleITK like this
```
git checkout 'v2.0.2'
mkdir build
cd build
cmake \
			-DCMAKE_INSTALL_PREFIX=/usr \
			-DCMAKE_CXX_FLAGS:STRING="-std=c++14" \
			-DLUA_VERSION_STRING:STRING="5.3.6" \
			-DLUA_EXECUTABLE:FILEPATH="/usr/bin/lua5.3" \
			-DLUA_INCLUDE_DIR:FILEPATH="/usr/include/lua5.3" \
			-DCMAKE_SKIP_RPATH:BOOL=ON \
			-DBUILD_SHARED_LIBS:BOOL=ON \
			-DBUILD_TESTING:BOOL=OFF \
			-DBUILD_EXAMPLES:BOOL=OFF \
			-DBUILD_DOXYGEN:BOOL=OFF \
			-DSimpleITK_PYTHON_WHEEL:BOOL=ON \
			-DWRAP_DEFAULT:BOOL=ON \
			-DWRAP_CSHARP:BOOL=ON \
			-DWRAP_JAVA:BOOL=ON \
			-DWRAP_LUA:BOOL=ON \
			-DWRAP_PYTHON:BOOL=ON \
			-DWRAP_R:BOOL=ON \
			-DWRAP_RUBY:BOOL=ON \
			-DWRAP_TCL:BOOL=ON \
			..
make all PythonVirtualEnv dist
```

Error for the Python wrapper:
```
created virtual environment CPython3.9.2.final.0-64 in 99ms
  creator CPython3Posix(dest=/home/martino/Desktop/simpleitk/src/SimpleITK/build/Testing/Installation/PythonVirtualenv, clear=True, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/martino/.local/share/virtualenv)
    added seed packages: setuptools==52.0.0, wheel==0.36.2
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator
-- Installing Python Package: SimpleITK
CMake Error at PythonVirtualEnvInstall.cmake:57 (message):
  Installation of SimpleITK into Python virutal enviroment failed.

  Traceback (most recent call last):

    File "/home/martino/Desktop/simpleitk/src/SimpleITK/build/Wrapping/Python/setup.py", line 14, in <module>
      from SimpleITK._version import __version__
    File "/home/martino/Desktop/simpleitk/src/SimpleITK/build/Wrapping/Python/SimpleITK/__init__.py", line 18, in <module>
      from SimpleITK.SimpleITK import *
    File "/home/martino/Desktop/simpleitk/src/SimpleITK/build/Wrapping/Python/SimpleITK/SimpleITK.py", line 13, in <module>
      from . import _SimpleITK

  ImportError: libSimpleITKBasicFilters1-2.0.so.1: cannot open shared object
  file: No such file or directory
```

Error for the TCL wrapper:
```
[ 99%] Linking CXX executable bin/SimpleITKTclsh
/usr/bin/ld: CMakeFiles/SimpleITKTclsh.dir/SimpleITKTCL_wrap.cxx.o: in function `_wrap_new_PasteImageFilter':
SimpleITKTCL_wrap.cxx:(.text+0xd14e9): undefined reference to `itk::simple::PasteImageFilter::PasteImageFilter()'
/usr/bin/ld: CMakeFiles/SimpleITKTclsh.dir/SimpleITKTCL_wrap.cxx.o: in function `_wrap_PasteImageFilter_Execute__SWIG_0':
SimpleITKTCL_wrap.cxx:(.text+0xd3d95): undefined reference to `itk::simple::PasteImageFilter::Execute(itk::simple::Image const&, itk::simple::Image const&)'
/usr/bin/ld: CMakeFiles/SimpleITKTclsh.dir/SimpleITKTCL_wrap.cxx.o: in function `_wrap_PasteImageFilter_Execute__SWIG_1':
SimpleITKTCL_wrap.cxx:(.text+0xd424a): undefined reference to `itk::simple::PasteImageFilter::Execute(itk::simple::Image const&, double)'
/usr/bin/ld: CMakeFiles/SimpleITKTclsh.dir/SimpleITKTCL_wrap.cxx.o: in function `_wrap_Paste__SWIG_0':
SimpleITKTCL_wrap.cxx:(.text+0xd4fd6): undefined reference to `itk::simple::Paste(itk::simple::Image const&, itk::simple::Image const&, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<int, std::allocator<int> >, std::vector<int, std::allocator<int> >, std::vector<bool, std::allocator<bool> >)'
/usr/bin/ld: CMakeFiles/SimpleITKTclsh.dir/SimpleITKTCL_wrap.cxx.o: in function `_wrap_Paste__SWIG_1':
SimpleITKTCL_wrap.cxx:(.text+0xd5a6c): undefined reference to `itk::simple::Paste(itk::simple::Image const&, itk::simple::Image const&, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<int, std::allocator<int> >, std::vector<int, std::allocator<int> >, std::vector<bool, std::allocator<bool> >)'
/usr/bin/ld: CMakeFiles/SimpleITKTclsh.dir/SimpleITKTCL_wrap.cxx.o: in function `_wrap_Paste__SWIG_2':
SimpleITKTCL_wrap.cxx:(.text+0xd6381): undefined reference to `itk::simple::Paste(itk::simple::Image const&, itk::simple::Image const&, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<int, std::allocator<int> >, std::vector<int, std::allocator<int> >, std::vector<bool, std::allocator<bool> >)'
/usr/bin/ld: CMakeFiles/SimpleITKTclsh.dir/SimpleITKTCL_wrap.cxx.o: in function `_wrap_Paste__SWIG_3':
SimpleITKTCL_wrap.cxx:(.text+0xd6b3d): undefined reference to `itk::simple::Paste(itk::simple::Image const&, itk::simple::Image const&, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<int, std::allocator<int> >, std::vector<int, std::allocator<int> >, std::vector<bool, std::allocator<bool> >)'
/usr/bin/ld: CMakeFiles/SimpleITKTclsh.dir/SimpleITKTCL_wrap.cxx.o: in function `_wrap_Paste__SWIG_4':
SimpleITKTCL_wrap.cxx:(.text+0xd7194): undefined reference to `itk::simple::Paste(itk::simple::Image const&, itk::simple::Image const&, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<int, std::allocator<int> >, std::vector<int, std::allocator<int> >, std::vector<bool, std::allocator<bool> >)'
collect2: fel: ld returnerade avslutningsstatus 1
```
Error for the R wrapper:
```
* installing *source* package ‘SimpleITK’ ...
** using staged installation
** libs
** arch - 
Dummy makefile for binary package
installing to /home/martino/Desktop/simpleitk/src/SimpleITK/build/Wrapping/R/R_libs/00LOCK-SimpleITK/00new/SimpleITK/libs
** R
** inst
** byte-compile and prepare package for lazy loading
Creating a generic function for ‘xor’ from package ‘base’ in package ‘SimpleITK’
** help
*** installing help indices
** building package indices
** installing vignettes
   ‘SimpleITK_tutorial.Rnw’ 
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘SimpleITK’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/home/martino/Desktop/simpleitk/src/SimpleITK/build/Wrapping/R/R_libs/00LOCK-SimpleITK/00new/SimpleITK/libs/SimpleITK.so':
  /home/martino/Desktop/simpleitk/src/SimpleITK/build/Wrapping/R/R_libs/00LOCK-SimpleITK/00new/SimpleITK/libs/SimpleITK.so: undefined symbol: _ZN3itk6simple16PasteImageFilter7ExecuteERKNS0_5ImageES4_
```

**Comment by blowekamp** [comment]
Thank you for working to get the library packaged.

This was an update from 1.x to 2.0.x? 

For R and Python, I am suspicious that a dependent library to the language specific libraries are specified with an embedded RPATH or in the runtime path. The libraries can be inspected with a tools such as `ldd` to see dependencies and if they include path.

Try setting the `LD_LIBRARY_PATH` env to the build output location of the SimpleITK libraries. It should be `<SIMPLEITK_BINARY_DIR>/lib`.

**Comment by m-pilia** [comment]
Thanks for the reply! Yes, I encountered this issue only when updating from 1.2.4 to 2.0.1. I tried fiddling with the RPATH settings in CMake and also to set LD_LIBRARY_PATH, but I still get the same issue. I also tried re-building dependencies such as ITK just to be sure. 
I will try to look a bit closer to the individual linker commands and missing symbols and see if I can figure out what I am missing, but so far I could not find any obvious explanation.

**Comment by blowekamp** [comment]
Another option to consider is to create a separate package for the "core" SimpleITK libraries. This library could be a dependency of the different wrapping languages packages. Since the SimpleITK "core" libraries would be a separate installed packaged, it would use the system/install policies for using the library. Let me know if you are interested in trying this approach and I can give you more information.

**Comment by m-pilia** [comment]
Sorry for the late reply, unfortunately I have been quite busy lately. Splitting into separate packages would be a good idea (that I would consider regardless). Would that require a different build configuration? One thing I had already tried was to build the and install the C++ library before building the wrappers, but it did not help.

**Comment by blowekamp** [comment]
It would involve configure, building, and packaging the root of the source with WRAP_DEFAULT=OFF. Then configuring, building and packaging the directories for each language under the Wrapping directory. Both of these build would require the same full source code.

What error did you encounter when just building the C++ library?

**Comment by m-pilia** [comment]
Sorry again for the late reply. I finally found some time for a deep dive into this issue.

I think the linker was right all along, and the problem was indeed the missing PasteImageFilter symbols in the SimpleITK shared libraries (I checked the build artifacts with `scanelf`). I then realised the `sitkPasteImageFilter.cxx` file was not being built at all, and after reading the `CMakeLists.txt` I noticed a dependency on the `SimpleITKFilters` ITK module (that is not part of a default ITK build, and is not included in the [insight-toolkit](https://aur.archlinux.org/packages/insight-toolkit) package on Arch). So I rebuilt ITK with the missing module, and the problem is solved.

My bad, if I had checked the flags used in `SuperBuild/External_ITK.cmake` I would have realised this from the beginning. Though it would probably help to get a CMake warning when trying to build SimpleITK with a missing ITK module :-)

Thanks a lot @blowekamp for the support.
